c/meterpreter/source/extensions/stdapi/server/net/socket/tcp_server.c

```diff
@@ -319,29 +382,46 @@ DWORD request_net_tcp_server_channel_open(Remote * remote, Packet * packet)
 				// support IPv4 and IPv6 we'd need to create two separate sockets. IPv6 on XP isn't that common
 				// so instead, we'll just revert back to v4 and listen on that one address instead.
 				closesocket(ctx->fd);
-				v4Fallback = TRUE;
+				ai_family = AF_INET;
+				ctx->fd = WSASocket(ai_family, SOCK_STREAM, IPPROTO_TCP, 0, 0, 0);
 			}
 		}
+		if (ctx->fd == INVALID_SOCKET) {
+			BREAK_ON_WSAERROR("[TCP-SERVER] request_net_tcp_server_channel_open. WSASocket failed");
+		}
 
 		struct sockaddr_in6 sockAddr = { 0 };
 		DWORD sockAddrSize = 0;
 
-		if (v4Fallback)
+		if (ai_family == AF_INET)
 		{
 			struct sockaddr_in* v4Addr = (struct sockaddr_in*)&sockAddr;
-			v4Addr->sin_addr.s_addr = localHost == NULL ? htons(INADDR_ANY) : inet_addr(localHost);
+			if (localHost) {
+				inet_pton(AF_INET, localHost, &v4Addr->sin_addr);

```

Comment:
This could fail and we aren't checking the result of this call in several places within this code.

Reply:
When would it fail? If the address is set, we've already validated that its either AF_INET or AF_INET6 which means we were able to decode it.

Reply:
When you are returning anything other than 1 from this function, you are encountering a failure case. The callers of this function should be checking if this function is returning 1 and if not they need to be gracefully failing vs proceeding to operate on data that is clearly not valid at that point.